### ISSUE TITLE
Single dispatch `issparse`

### ISSUE THREAD

**Issue by ivirshup** [scipy/scipy#11565]
**Is your feature request related to a problem? Please describe.**

I would like a single way to tell if a matrix or array is sparse. Right now it's `isinstance(a, (*types_that_implement_sparse_arrays_or_matrices))`

**Describe the solution you'd like**

```python
from functools import singledispatch

@singledispatch
def issparse(x):
    return False

@issparse.register(scipy.sparse.spmatrix)
def _issparse_spmatrix(x):
    return True
```

This way, other packages can import `issparse` and overload it on their types. I.e. `pydata/sparse` could have:

```python
from scipy.sparse import issparse

@issparse.register(sparse.SparseArray)
def _issparse_sparsearray(x):
    return True 
```

**Describe alternatives you've considered**

* `isinstance`, but now I have to explicitly list all the types this could be. Additionally I can't catch types from a new package.

**Additional context (e.g. screenshots)**

This is similar to how the Julia community does traits, which I've found quite useful.

**Comment by ivirshup** [comment]
@hameerabbasi, I would like to get your opinion on this idea. Especially since it would only be useful if it was also implemented by `pydata/sparse`.

**Comment by hameerabbasi** [comment]
I think the right way to do this in Python would be protocols. For example, a `__csr__` protocol, or similar.

**Comment by ivirshup** [comment]
Would you mind elaborating on that a bit?

* What would the `__csr__` attribute be? Is it a method, or a value? And what does it tell us about the object?
* Would you want this for every kind of sparse format? Will many libraries be implementing the same sparse formats?
    * Subquestion, will pydata/sparse have `CSR` and `CSC` classes?
* Could there be a catch-all check for sparse arrays?

--------------

I thought I'd expand on this with an example use-case. When I implement a method for sparse data, I've probably only defined it for one sparse format – one which matches the access pattern. So my code frequently looks like:

```python
def some_method(A):
    if issparse(A):
        if not isinstance(A, csr_matrix):
            A = csr_matrix(A)  # convert format (this could use an extendable interface too?)
        return _some_method_sparse(A)
    else:
        return _some_method_dense(A)
```

It would be nice if this pattern worked across sparse array packages.

**Comment by hameerabbasi** [comment]
Okay, I thought some more about this. Since PyData/Sparse will be the standard moving forward, there should be a `__sparsearray__` method on all objects that can convert to sparse arrays that PyData/Sparse supports, doesn't matter which format. This is similar to how `__array__` works for dense arrays: It returns a dense NumPy array or something following the buffer protocol.

So, for example, in your code, you would simply do the following (as an example):

```python
def some_function(x):
    try:
        return sparse_function(x)
    except TypeError:
        return dense_function(x)
```

Does that make sense?

Another thing we could do in theory is to allow dense passthrough in sparse functions, but something tells me that's going to blow up in many cases.

**Comment by ivirshup** [comment]
I'm not quite sure. A dense numpy array could be converted to a sparse array, but would dense arrays get a `__sparsearray__` method? Sparse formats can be very different, is the object converting itself to an arbitrary one or is the user specifying?

> I'm working on a blog post regarding future work,

That does sound very cool, especially if I would be able to define my own semirings and have the compiler take care of data access. Looking forward to reading that!

**Comment by hameerabbasi** [comment]
> I'm not quite sure. A dense numpy array could be converted to a sparse array, but would dense arrays get a `__sparsearray__` method?

No, they won't. Even if they did, it'd return `NotImplemented`.
